internal/ls/utilities.go

```diff
@@ -39,6 +41,13 @@ func CompareRanges(lsRange, other *lsproto.Range) int {
 	return ComparePositions(lsRange.End, other.End)
 }
 
+func applyTextChanges(text string, changes []core.TextChange) string {
+	for i := len(changes) - 1; i >= 0; i-- {
+		text = changes[i].ApplyTo(text)
```

Comment:
When there are many edits (like, say, 30000 to re-indent every line of `checker.ts`), this algorithm produces a _huge_ quantity of trash strings (gigabytes worth). There's an `applyBulkEdits` function in `tsc.go` that's already worked around the issue by avoiding individual `.ApplyTo` calls. You'll probably want to move it and use it in the LS, too.